### PR TITLE
Support more arguments to tail calls by passing them through the domain state

### PR DESCRIPTION
**Pull request by xavierleroy**
In 2004, commit af9b98fcb, the calling conventions for the i386 port of ocamlopt were changed: the first 6 integer arguments go into registers, like before, but the next 16 arguments go into a global array `caml_extra_params`, instead of being passed on stack like before.  The reason for this hack is that passing arguments in global memory does not preclude tail call optimization, unlike passing arguments on stack.  Parameters passed via `caml_extra_params` are immediately copied on stack or in registers on function entry, before another function call, a GC, or a context switch can take place, so everything is safe, as long as true parallelism is not supported.

This hack was justified by the paucity of registers provided by the i386 architecture.  It was believed that other architectures provide enough registers for parameter passing that most if not all reasonable tail calls can be accommodated.

Now it's 2021 and users want tail calls with more arguments than available registers on all the architectures we support.

So, biting the bullet and swallowing some pride, this commit extends the 2004 i386 hack to all the architectures supported by OCaml.  Once the registers available for passing function arguments are exhausted, the next 64 arguments are passed in a memory area that is part of the domain state.  This argument passing is compatible with tail calls, so we get guaranteed tail calls up to 70 arguments (in the worst case).

The domain state is used instead of a global array so that (1) this is compatible with Multicore OCaml and parallel execution of multiple domains, and (2) we benefit from efficient addressing from the domain state register.

For i386, we don't have a domain state register, and Multicore OCaml will support only one domain on this architecture, so we keep using a global `caml_extra_params` array; only, its size was increased to accommodate 64 arguments.

### CODE REVIEW

**Comment by gasche** [comment]
Thanks! This is very reassuring for the TRMC work (which introduces 2 extra arguments (possibly 3 in the future) per transformed function, making us nervous about breaking tail-call-ness). And the trick somehow feels more noble when it is used pervasively rather than just for one architecture.

**Comment by gasche** [comment]
I read the code. It seems okay to me, but I don't know much about the backend so I don't think I can approve meaningfully.

I'm not fond of the way `Domainstate` is handled as a special snowflake in the emit functions (it's part of a type of "slots" but sometimes the functions expecting slots don't support it); I have the impression that the code would be nicer if the `slot_offset` functions turned into `slot_address`, returning both "base" and "offset" components. (`emit_store` and `emit_load` helper functions, on backends that have them, could take the `(base, offset)` pair directly as an argument.)

File: asmcomp/arm/proc.ml [hunk 1]
```diff
@@ -109,6 +109,8 @@ let stack_slot slot ty =
 
 (* Calling conventions *)
 
+let size_domainstate_args = 64 * 8
```
Comment (lthls) [inline asmcomp/arm/proc.ml]:
```suggestion
let size_domainstate_args = 64 * 4
```

File: asmcomp/arm/emit.mlp [hunk 2]
```diff
@@ -74,13 +74,18 @@ let slot_offset env loc cl =
...
       assert (n >= 0);
       n
+  | Domainstate _ -> assert false  (* not a stack slot *)
 
 (* Output a stack reference *)
 
 let emit_stack env r =
   match r.loc with
+  | Stack (Domainstate n) ->
+      let ofs = n + Domainstate.(idx_of_field Domain_extra_params) * 8 in
```
Comment (lthls) [inline asmcomp/arm/emit.mlp:84]:
```suggestion
      let ofs = n + Domainstate.(idx_of_field Domain_extra_params) * 4 in
```

File: asmcomp/arm/proc.ml (hunk 1 above)
Comment (xavierleroy) [reply asmcomp/arm/proc.ml]: Well spotted.  I put `64 * size_int` for all ports, it's clearer and less error prone.

File: asmcomp/arm/emit.mlp (hunk 2 above)
Comment (xavierleroy) [reply asmcomp/arm/emit.mlp:84]: No. Even on 32-bit platforms, entries in the domain state are 8 bytes wide.  

Note for later: introduce a `Domainstate.offset_of_field` function that does the `* 8`.

**Comment by xavierleroy** [comment]
> I have the impression that the code would be nicer if the slot_offset functions turned into slot_address, returning both "base" and "offset" components. (emit_store and emit_load helper functions, on backends that have them, could take the (base, offset) pair directly as an argument.)

The RISC-V port follows this approach, because register-indirect addressing is more complicated on RISC-V than on the other ports.  One downside is that `record_frame_label` then needs to assert that the base register is the stack pointer, because Domainstate slots cannot be GC roots.  On other ports, the assertion is part of `slot_offset`.

File: asmcomp/arm/emit.mlp (hunk 2 above)
Comment (lthls) [reply asmcomp/arm/emit.mlp:84]: Ah, of course. Somewhere in the middle of reviewing I thought this was actually computing an offset in the `extra_params` array.

File: asmcomp/power/proc.ml [hunk 3]
```diff
@@ -136,36 +138,43 @@ let loc_int_pair last_int make_stack int ofs =
...
 
 let loc_parameters arg =
-  let (loc, _ofs) = calling_conventions 0 15 100 112 incoming arg
+  let (loc, _ofs) =
+    calling_conventions 0 15 100 112 incoming (- size_domainstate_args) arg
   in loc
 
 let loc_results res =
-  let (loc, _ofs) = calling_conventions 0 15 100 112 not_supported res
+  let (loc, _ofs) = calling_conventions 0 7 100 112 not_supported 0 res
```
Comment (lthls) [inline asmcomp/power/proc.ml]: There seems to be an accidental reversal to 8 integer registers.

File: asmcomp/power/emit.mlp [hunk 4]
```diff
@@ -133,8 +134,12 @@ let emit_reg r =
 
 let emit_stack env r =
   match r.loc with
+  | Stack (Domainstate n) ->
+      let ofs = n + Domainstate.(idx_of_field Domain_extra_params) * 8 in
+      `{emit_int ofs}(30)`
```
Comment (lthls) [inline asmcomp/power/emit.mlp:139]: The comment at the start of `proc.ml` associates the domain state pointer to register `28`.

File: asmcomp/riscv/emit.mlp [hunk 5]
```diff
@@ -39,12 +39,17 @@ let frame_size env =
...
+       if cls = 0
+       then env.stack_offset + env.f.fun_num_stack_slots.(1) * size_float
+            + n * size_int
+       else env.stack_offset + n * size_float)
+  | Incoming n ->
+      ("sp", frame_size env + n)
+  | Outgoing n ->
+      ("sp", n)
+  | Domainstate n -> 
+      ("s11", n + Domainstate.(idx_of_field Domain_extra_params) * 8)
```
Comment (lthls) [inline asmcomp/riscv/emit.mlp:52]: The comment in `proc.ml` points to `"s0"` as the register holding the domain state pointer.

File: asmcomp/power/proc.ml (hunk 3 above)
Comment (xavierleroy) [reply asmcomp/power/proc.ml]: Yes, a rebase from trunk is in order...

File: asmcomp/power/emit.mlp (hunk 4 above)
Comment (xavierleroy) [reply asmcomp/power/emit.mlp:139]: That's not what I see:
https://github.com/ocaml/ocaml/blob/4e9eb755e60d0f0f4e458eb3f7af26bce5d8e2ba/asmcomp/power/proc.ml#L39

File: asmcomp/riscv/emit.mlp (hunk 5 above)
Comment (xavierleroy) [reply asmcomp/riscv/emit.mlp:52]: That's not what I see: https://github.com/ocaml/ocaml/blob/4e9eb755e60d0f0f4e458eb3f7af26bce5d8e2ba/asmcomp/riscv/proc.ml#L44

File: asmcomp/power/proc.ml (hunk 3 above)
Comment (xavierleroy) [reply asmcomp/power/proc.ml]: Should be fixed now.  Thanks for spotting this.

File: asmcomp/riscv/emit.mlp (hunk 5 above)
Comment (lthls) [reply asmcomp/riscv/emit.mlp:52]: Sorry, I must have been looking at an older version (I'm working on too many versions at once...).

**Comment by xavierleroy** [comment]
Something was missing in the RISC-V port.  Now, all architectures pass (CI precheck + manual RISC-V build).

**Review by lthls — APPROVED** [review]
I've gone through all the code and reported everything I found suspect (and I was only wrong about half of the time).
I've only looked at the changes to the runtime in relation to the changes in the backend though. I don't know if something is going to break because the last field of the domain state hasn't the same size as the other fields anymore (but I trust @xavierleroy knows what he's doing).

The only nitpick I have left: the `make_stack` argument to `calling_conventions` is a bit misleading, as it does not always allocate on the stack. But I don't have a good alternative (`make_slot` ? `make_extra_slot` ?).

**Comment by xavierleroy** [comment]
Thanks @lthls for the review.  I'm thinking about adding some specific tests, so I'll wait a bit more before merging.

> The only nitpick I have left: the make_stack argument to calling_conventions is a bit misleading, as it does not always allocate on the stack. But I don't have a good alternative (make_slot ? make_extra_slot ?).

The misleading naming starts with the `Stack` constructor of type `Reg.location` and its argument of type `Reg.stack_location` that now includes memory locations that are not in the stack...  We could use "mem" instead of "stack", e.g. in reg.mli
```
and location =
    Unknown
  | Reg of int
  | Mem of memory_location

and memory_location =
    Local of int
  | Incoming of int
  | Outgoing of int
  | Domainstate of int
```
Then, the `make_stack` arguments to `calling_conventions` could be renamed `make_mem`.  But I don't know how much code (e.g. in Flambda2...) we're breaking with this renaming.

**Comment by xavierleroy** [comment]
> I don't know if something is going to break because the last field of the domain state hasn't the same size as the other fields anymore

I strongly believe we'll be fine as long as it's only the last field that has size bigger than 8.  If we need multiple fields of size > 8, we'll need different, more complex macro-processing of `runtime/caml/domain_state.tbl`.

**Comment by EduardoRFS** [comment]
Probably dumb question, why can't we have tail call recursion with stack arguments? Isn't it a matter of just writing the arguments ignoring the top element of the stack?

**Comment by gasche** [comment]
@EduardoRFS this was discussed in https://github.com/ocaml/ocaml/pull/10426#issuecomment-847854334. In summary, the difficulty is making the runtime/GC aware that this stack fragment changes ownership from the caller to the callee.

**Comment by xavierleroy** [comment]
> the difficulty is making the runtime/GC aware that this stack fragment changes ownership from the caller to the callee.

Right.  I thought some more about it and now I think it could be done, but it would require a more complex format for stack frame descriptors.

Also, if tailcall arguments were passed on stack, they could occupy the same location as local stack slots of the caller, so the caller would either need a "parallel assignment" algorithm, with enough temporary locations to avoid overwriting local slots that are still needed, or more brutally, set up the arguments at the bottom of the caller's frame, like for a non-tail call, then copy them over to their final destination.  All this is more complicated and slower (at run-time) than going through the domain state locations.

**Comment by mshinwell** [comment]
@xavierleroy Absolutely fine from the Flambda 2 side if you want to do the proposed renaming (which seems like a good idea) -- we will adapt.

**Comment by xavierleroy** [comment]
> Absolutely fine from the Flambda 2 side if you want to do the proposed renaming (which seems like a good idea) -- we will adapt.

Good to know, thanks.  I'll leave the renaming for a future PR and merge this PR in its present state, so that it gets tested in the wild.

**Pull request merged**